From d63f037b70e0fe2e48f3f1f24b28869d890cc0f9 Mon Sep 17 00:00:00 2001
From: Shunqian Zheng <zhengsq@rock-chips.com>
Date: Mon, 19 Aug 2019 07:06:56 +0000
Subject: [PATCH] media: rockchip: rkisp1: restrict video output resolution

Currently the maximum size of video is 4416x3312. For those sensors(5M
or 8M), it's not a good idea to scale up to max resolution by ISP.

This avoid video output resolution exceeding the isp sub-device
size.

Change-Id: I4b2649cf3a8c3c7e81ff978eefba3e69e2c24d81
Signed-off-by: Shunqian Zheng <zhengsq@rock-chips.com>
---
 .../media/platform/rockchip/isp1/capture.c    | 26 ++++++++++++++++---
 1 file changed, 22 insertions(+), 4 deletions(-)

diff --git a/drivers/media/platform/rockchip/isp1/capture.c b/drivers/media/platform/rockchip/isp1/capture.c
index a362e4243c5d..88072b19cbe8 100644
--- a/drivers/media/platform/rockchip/isp1/capture.c
+++ b/drivers/media/platform/rockchip/isp1/capture.c
@@ -1600,6 +1600,23 @@ static int rkisp_init_vb2_queue(struct vb2_queue *q,
 	return vb2_queue_init(q);
 }
 
+/*
+ * Make sure max resize/output resolution is smaller than
+ * isp sub device output size. This assumes it's not
+ * recommended to use ISP scale-up function to get output size
+ * that exceeds sensor max resolution.
+ */
+static void restrict_rsz_resolution(struct rkisp1_device *dev,
+				    const struct stream_config *config,
+				    struct v4l2_rect *max_rsz)
+{
+	struct v4l2_rect *input_win;
+
+	input_win = rkisp1_get_isp_sd_win(&dev->isp_sdev);
+	max_rsz->width = min_t(int, input_win->width, config->max_rsz_width);
+	max_rsz->height = min_t(int, input_win->height, config->max_rsz_height);
+}
+
 static int rkisp1_set_fmt(struct rkisp1_stream *stream,
 			   struct v4l2_pix_format_mplane *pixm,
 			   bool try)
@@ -1624,15 +1641,16 @@ static int rkisp1_set_fmt(struct rkisp1_stream *stream,
 	}
 
 	if (stream->id != RKISP1_STREAM_RAW) {
+		struct v4l2_rect max_rsz;
+
 		other_stream =
 			&stream->ispdev->stream[!stream->id ^ 1];
 		/* do checks on resolution */
+		restrict_rsz_resolution(stream->ispdev, config, &max_rsz);
 		pixm->width = clamp_t(u32, pixm->width,
-			config->min_rsz_width,
-			config->max_rsz_width);
+				      config->min_rsz_width, max_rsz.width);
 		pixm->height = clamp_t(u32, pixm->height,
-			config->min_rsz_height,
-			config->max_rsz_height);
+				       config->min_rsz_height, max_rsz.height);
 	} else {
 		other_stream =
 			&stream->ispdev->stream[RKISP1_STREAM_MP];
-- 
2.35.3

